elegant/worm_data.py

```diff
@@ -441,6 +441,78 @@ def interpolate_feature(self, feature, age, age_feature='age'):
         ages, value = self.get_time_range(feature, age_feature=age_feature)
         return numpy.interp(age, ages, value)
 
+    def make_smoothing_filter(filter_type, **filter_params):
+        '''Generates a causal smoothing filter (i.e. one that only smooths over past values in data)
+
+            Parameters
+                filter_type - str corresponding to the filter to be made ('gaussian', 'uniform')
+                **filter_params - parameters for the specified filter; these include:
+
+                    window_length - required parameter for either filter (used for the gaussian to truncate it)
+                    sigma - used only for gaussian filters
+
+            Returns
+                The generated filter will have call signature filter(time, data)
+                where: (1) time has been previously centered so that t=0 represents
+                the current timepoint and positive values of time correspond to past data,
+                and (2) data contains values to smooth over.
+        '''
+
+        def gaussian_filter(time, data, window_length, sigma):
+            smoothing_mask = (time>=0) & (time <= window_length)
+            t = time[smoothing_mask]
+            return numpy.dot(1/numpy.exp(-t**2/(2*sigma)).sum() * numpy.exp(-t**2/(2*sigma)), data[smoothing_mask])
+        def uniform_filter(time, data, window_length):
+            smoothing_mask = (time>=0) & (time <= window_length)
+            return numpy.dot(smoothing_mask/smoothing_mask.sum(),data)
+
+        filters = {'gaussian': gaussian_filter,
+            'uniform': uniform_filter}
+
+        def smoothing_filter(time, data):
+            selected_filter = filters[filter_type](**filter_params)
+            return selected_filter(time,data)
+        return smoothing_filter
+
+
+    def smooth_feature(self, feature, filter_type, age_feature='age', min_age=-numpy.inf, max_age=numpy.inf, **filter_params):
+        """Smooth a feature with a prescribed filter
+
+            Parameters
+                feature - feature to filter over
+                filter_type - str corresponding to the filter to be applied ('gaussian', 'uniform'),
+                    or a callable with the call signature smoothing_filter(time,data)
+                age_feature: the name of the feature to use to determine the age
+                    window. Defaults to 'age', but other monotonic features, such as
+                    a time-left-alive 'ghost_age' could also be used. If this is a
+                    function, it will be called as age_feature(worm) to generate
```

Comment:
There's a problem with `age_feature` being a function: `get_time_range` does the right thing, but later on you assume it's an attribute. Either get all the ages as `ages = age_feature(self) if callable(age_feature) else getattr(self.td, age_feature)` or don't advertise the possibility of it being callable.

My suspicion is that `get_time_range` shouldn't allow a callable either -- that just gets too confusing. I'll probably strip that out later.